cdc/sinkv2/eventsink/mq/producer/kafka_producer.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package producer
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/contextutil"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sink/mq/codec"
+	"github.com/pingcap/tiflow/cdc/sinkv2/eventsink"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	"github.com/pingcap/tiflow/pkg/util"
+	"go.uber.org/zap"
+)
+
+var _ Producer = (*kafkaProducer)(nil)
+
+// messageMetaData is used to store the callback function for the message.
+type messageMetaData struct {
+	callback eventsink.CallbackFunc
+}
+
+// kafkaProducer is used to send messages to kafka.
+type kafkaProducer struct {
+	// id indicates this sink belongs to which processor(changefeed).
+	id model.ChangeFeedID
+	// role indicates this sink used for what.
+	role util.Role
+	// We hold the client to make close operation faster.
+	// Please see the comment of Close().
+	client sarama.Client
+	// asyncProducer is used to send messages to kafka asynchronously.
+	asyncProducer sarama.AsyncProducer
+	// closedMu is used to protect `closed`.
+	// We need to ensure that never write to closed producers.
+	closedMu sync.RWMutex
+	// closed is used to indicate whether the producer is closed.
+	// We also use it to guard against double closed.
+	closed bool
+	// closedChan is used to notify the run loop to exit.
+	closedChan chan struct{}
+	// failpointCh is used to notify the run loop to failpoint.
+	// Only used in test.
+	failpointCh chan error
+}
+
+func (k *kafkaProducer) AsyncSendMessage(
+	ctx context.Context, topic string,
+	partition int32, message *codec.MQMessage,
+) error {
+	// We have to hold the lock to prevent write to closed producer.
+	// Close may be blocked for a long time.
+	k.closedMu.RLock()
+	defer k.closedMu.RUnlock()
+
+	// If the producer is closed, we should skip the message.
+	if k.closed {
+		// We return the context error directly rather than other error.
+		// Because if producer already closed, it means the context maybe is canceling or canceled.
+		// So we shouldn't return other errors to the caller to mess up the shutdown process.
+		return ctx.Err()
```

Comment:
I think returning a special error for `ProducedClosed` may be better.

Reply:
I was afraid that returning a non-cancel error would lead to strange behavior, since the original implementation didn't return an error directly.

Reply:
I still think returning `ctx.Err()` introduces additional coupling, making error handling more complicated in the long term.

Reply:
OK, I will remove it.

Reply:
Please checkout https://github.com/pingcap/tiflow/pull/6412/commits/1f7ea637d619cc8dafe0e325ad6bde515da26983